json.go

```diff
@@ -0,0 +1,809 @@
+	case uint32:
+		C.godror_dpiJson_setUint64(jsonnode, C.uint64_t(x))
+	case uint64:
+		C.godror_dpiJson_setUint64(jsonnode, C.uint64_t(x))
+	case float32:
+		C.godror_dpiJson_setDouble(jsonnode, C.double(x))
+	case float64:
+		C.godror_dpiJson_setDouble(jsonnode, C.double(x))
+	case string:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setString(jsonnode, &(data.dpiData))
+	case time.Time:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setTime(jsonnode, &(data.dpiData))
+	case time.Duration:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setIntervalDS(jsonnode, &(data.dpiData))
+	case []byte:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setBytes(jsonnode, &(data.dpiData))
+	case bool:
+		data, err := NewData(x)
+		if err != nil {
+			return err
+		}
+		C.godror_dpiJson_setBool(jsonnode, &(data.dpiData))
+	default:
+		return fmt.Errorf("Unsupported type %T\n", in)
+	}
+	return nil
+}
+
+// Creates a JSONInt from user input []interface{}
+func newJSONScalar(val interface{}, jscalar *JSONScalar) error {
+	var dpijsonnode *C.dpiJsonNode
+	C.godror_allocate_dpiNode((**C.dpiJsonNode)(unsafe.Pointer(&dpijsonnode)))
+	err := populateJsonNode(val, dpijsonnode)
+	if err != nil {
+		C.godror_dpiJsonfreeMem(dpijsonnode)
+		return err
+	}
+	dpidataw := new(Data)
+	jsonNodeToData(dpidataw, dpijsonnode)
+	*jscalar = JSONScalar{dpiJsonNode: dpijsonnode, isMemOwned: true}
+	return nil
+}
+
+// Creates a JSONArray from user input []interface{}
+func newJSONArray(arr []interface{}, jsarr *JSONArray) error {
+	var dpijsonnode *C.dpiJsonNode
+	C.godror_allocate_dpiNode((**C.dpiJsonNode)(unsafe.Pointer(&dpijsonnode)))
+	err := populateJsonNode(arr, dpijsonnode)
+	if err != nil {
+		C.godror_dpiJsonfreeMem(dpijsonnode)
+		return err
+	}
+	dpidataw := new(Data)
+	jsonNodeToData(dpidataw, dpijsonnode)
+	*jsarr = JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(dpidataw.dpiData)), dpiJsonNode: dpijsonnode}
+	return nil
+}
+
+// Creates a JSONObject from user input map[string]interface{}
+func newJSONObject(m map[string]interface{}, jsobj *JSONObject) error {
+	var dpijsonnode *C.dpiJsonNode
+	C.godror_allocate_dpiNode((**C.dpiJsonNode)(unsafe.Pointer(&dpijsonnode)))
+	err := populateJsonNode(m, dpijsonnode)
+	if err != nil {
+		C.godror_dpiJsonfreeMem(dpijsonnode)
+		return err
+	}
+	dpidataw := new(Data)
+	jsonNodeToData(dpidataw, dpijsonnode)
+	*jsobj = JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(dpidataw.dpiData)), dpiJsonNode: dpijsonnode, isMemOwned: true}
+	return nil
+}
+
+// Frees the C memory associated with JSONObject
+func (jsobj *JSONObject) Close() error {
+	C.godror_dpiJsonfreeMem(jsobj.dpiJsonNode)
+	return nil
+}
+
+func (j JSONObject) Len() int { return int(j.dpiJsonObject.numFields) }
+func (j JSONObject) Get(m map[string]Data) {
+	n := int(j.dpiJsonObject.numFields)
```

Comment:
Why is the map an argument? Why not just return it?

Reply:
made it as return, took changes from your private json branch.